treewide: reduce the usage of iostream in headers

tchaikov opened the pull request:
for better compilation speed, as `<iostream>` is the one of the heaviest C++ headers.

in this series:

- stop using iostream in `seastar/net/packet-util.hh`
- extract `seastar::format()` into `format.hh`
- use `format.hh` directly

tested on my laptop with 32 G memory, 8c16t. 

- ccache and distcc are not enabled
- all interaction with the computer are stopped
- most of the active foreground tasks showed by `top` are stopped

before this change:
---


```console
$ git clean -dfx
$ ./configure.py --mode debug --disable-dpdk --compiler clang++      
$ time cmake --build build/debug
...
real    2m43.926s
user    35m58.795s
sys     1m50.180s
```

after this change:
---

```console
$ git clean -dfx
$ ./configure.py --mode debug --disable-dpdk --compiler clang++      
$ time cmake --build build/debug
...
real    2m42.688s
user    35m55.328s
sys     1m50.926s
```

i repeated the test for 3 times. the change shortened the build time 3 out of 3 times.

nyh commented:
> for better compilation speed, as `<iostream>` is the one of the heaviest C++ headers.

Would be nice to have such a statement be accompanied by a measurements that it really did improve compilation speed, or lowered object-file size (which empirically is usually proportional to the object file size), or an explanation why it might improve compilation speed in the future but doesn't yet because of I-don't-know-what`.

> in this series:
> 
>     * replace iostream with fprintf() when appropriate

Is it ever appropriate?
As far as I know, the C++ standard guarantees that C++ iostreams are "synced" with C stdio (use the same buffer, etc.) only for the standard streams (std::cout et al.), not for other files.
Also quick grep shows that today we don't have a even a single printf() call in Scylla. Do we really want to start using it?

A better approach is usually to move the output commands from the header file (.hh) to a source file (.cc), so that no matter how slow the output code is to compile - it's only compiled once.

avikivity commented on include/seastar/net/packet-util.hh:
```diff
@@ -145,7 +145,8 @@ public:
                 // If we reach here, we have a bug with merge.
-                std::cerr << "packet_merger: merge error\n";
+                std::fprintf(stderr, "packet_merger: merge error\n");
+                std::fflush(stderr);
```
Agree. An instead of switching to fprintf, we can just deinline code snippets that print something, if we think printing is needed.

tchaikov replied:
i replaced them with `on_internal_error()`, hopefully it should be better now.

avikivity commented on src/net/native-stack-impl.hh:
```diff
@@ -244,7 +245,8 @@ native_connected_socket_impl<Protocol>::get_nodelay() const {
     // FIXME: implement
-    std::cerr << "Keepalive is not supported by native stack" << std::endl;
+    std::fprintf(stderr, "Keepalive is not supported by native stack\n");
+    std::fflush(stderr);
```
You can use seastar_logger for these things.

tchaikov replied:
yeah, good idea. applied.

tchaikov commented on include/seastar/util/optimized_optional.hh:
```diff
@@ -23,7 +23,7 @@
 #include <seastar/util/modules.hh>
 #ifndef SEASTAR_MODULE
-#include <iostream>
+#include <cstdio>
```
sorry, it was a mechanical replace. i've reverted in the latest revision.

tchaikov commented on include/seastar/core/print.hh:
```diff
@@ -25,7 +25,6 @@
 #include <seastar/core/sstring.hh>
 #include <seastar/util/modules.hh>
 #ifndef SEASTAR_MODULE
-#include <fmt/ostream.h>
```
it's used by `::fmt::fprintf()`. which is in turn used by `fprint(std::ostream& os, const char* fmt, A&&... a)`. this function is marked deprecated. probably we can remove it in a follow-up change. but this is not in the scope of this changeset.

tchaikov commented:
> > for better compilation speed, as `<iostream>` is the one of the heaviest C++ headers.
> 
> Would be nice to have such a statement be accompanied by a measurements that it really did improve compilation speed, or lowered object-file size (which empirically is usually proportional to the object file size), or an explanation why it might improve compilation speed in the future but doesn't yet because of I-don't-know-what`.

- tested on my laptop with 32 G memory, 8c16t. 
- i was not using ccache or distcc. and
- stopped all interaction with the computer. 
- stopped most of the active foreground tasks showed by `top`

before this change:
---


```console
$ git clean -dfx
$ ./configure.py --mode debug --disable-dpdk --compiler clang++      
$ time cmake --build build/debug
...
real    2m44.731s
user    36m20.322s
sys     1m52.140s
```

after this change:
---

```console
$ git clean -dfx
$ ./configure.py --mode debug --disable-dpdk --compiler clang++      
$ time cmake --build build/debug
...
real    2m42.039s
user    35m50.410s
sys     1m50.728s
```

i repeated the test for 3 times. the change shortened the build time 3 out of 3 times.

tchaikov commented:
> an explanation why it might improve compilation speed in the future but doesn't yet because of I-don't-know-what`.

added a test result. but i think i've explained this:

> as `<iostream>` is the one of the heaviest C++ headers.

i think this is also why `ios_fwd` is provided by the standard library. but i am not able to provided a report for supporting this at this moment. but trading a relatively large header file with a much smaller one should be pure win, am i right?

v2:

- added the test results in the cover letter.
- dropped the print statement in the header. turns out we always `abort()` in other places in seastar. 
- reverted the change in `src/net/native-stack-impl.hh`, which is not a public-facing header. and is only included by .cc files.
- reverted the change in `include/seastar/util/optimized_optional.hh`. we should deprecate the operator<< overload in favor of formatter specialization. but let's leave it for another change.
